workspaces/liferay-dlfoldertemplate-workspace/client-extensions/liferay-dlfoldertemplate-custom-element/src/components/template-diagram/Diagram.js

```diff
@@ -0,0 +1,505 @@
+import ReactFlow, {
+	Background,
+	ConnectionLineType,
+	Controls,
+	Panel,
+	useEdgesState,
+	useNodesState,
+} from 'reactflow';
+
+import 'reactflow/dist/style.css';
+import ClayButton, {ClayButtonWithIcon} from '@clayui/button';
+import {ClayInput} from '@clayui/form';
+import {Form, Input} from 'antd';
+
+import {
+	addNode,
+	deleteFolderTemplateBatch,
+	getAvailableTemplatesNodesPage,
+	updateFolderTemplate,
+} from '../../services/TemplateDiagramService';
+import {showError} from '../../utils/util';
+import FolderNode from './controls/custom-node/FolderNode';
+
+import './Diagram.css';
+
+const dagreGraph = new dagre.graphlib.Graph();
+
+dagreGraph.setDefaultEdgeLabel(() => ({}));
+
+const nodeWidth = 172;
+
+const nodeHeight = 36;
+
+const position = {x: 0, y: 0};
+
+const edgeType = 'SmoothStep';
+
+const getLayoutedElements = (nodes, edges, direction = 'TB') => {
+	dagreGraph.setGraph({
+		rankdir: direction,
+	});
+
+	nodes.forEach((node) => {
+		dagreGraph.setNode(node.id, {height: nodeHeight, width: nodeWidth});
+	});
+
+	edges.forEach((edge) => {
+		dagreGraph.setEdge(edge.source, edge.target);
+	});
+
+	dagre.layout(dagreGraph);
+	nodes.forEach((node) => {
+		const nodeWithPosition = dagreGraph.node(node.id);
+
+		node.targetPosition = 'top';
+
+		node.sourcePosition = 'bottom';
+
+		node.draggable = false;
+
+		node.position = {
+			x: nodeWithPosition.x - nodeWidth / 2,
+
+			y: nodeWithPosition.y - nodeHeight / 2,
+		};
+
+		return node;
+	});
+
+	return {edges, nodes};
+};
+
+const getNodesFromHeadless = async (templateId) => {
+	const dBNodesPage = await getAvailableTemplatesNodesPage(templateId);
+
+	const dBNodes = dBNodesPage.items;
+
+	if (!dBNodes.length) {
+		const root = await addNode(0, true, 'Root Node', templateId);
+		dBNodes.push(root);
+	}
+
+	const dbNormalizedNodes = dBNodes.map((node) => {
+		return {
+			data: {
+				description: node.description,
+				label: node.name,
+				nodeId: `${node.id}`,
+				parent: node.root ? null : `${node.parentID}`,
+				root: node.root,
+			},
+			deletable: !node.root,
+			id: `${node.id}`,
+			parent: node.root ? null : `${node.parentID}`,
+			position,
+			type: node.root ? 'folderNode' : 'folderNode',
+		};
+	});
+
+	const dBEdges = [];
+
+	dBNodes.forEach((node) => {
+		if (node.parentID !== 0) {
+			const edge = {
+				animated: false,
+				id: `e${node.id}${node.parentID}`,
+				source: `${node.parentID}`,
+				target: `${node.id}`,
+				type: edgeType,
+			};
+			dBEdges.push(edge);
+		}
+	});
+
+	return getLayoutedElements(dbNormalizedNodes, dBEdges);
+};
+
+const getSubNodes = (parentNodeId) => {
+	const subNodes = [];
+
+	const checkSubNodes = (parentNode) => {
+		const childrenNodes = dagreGraph.successors(parentNode);
+
+		subNodes.push(parentNode);
+
+		if (childrenNodes && childrenNodes.length) {
+			childrenNodes.forEach((node) => {
+				checkSubNodes(node);
+			});
+		}
+	};
+
+	checkSubNodes(parentNodeId);
```

Comment:
Instead the whole function can be recursively called:
```
const getChildNodeIds(nodeId) {
  let childNodeIds = [];

  const successorIds = dagreGraph.successors(nodeId);

  if (successorIds && successorIds.length) {
    successorIds.forEach(successorId => {
      childNodeIds = [
        ...childNodeIds,
        successorId,
        getChildNodeIds(successorId),
      ];
    }
  }

  return childNodesIds;
}
```

The one difference is this function does not return the parentNodeId in the list. That could be done, but it would need to be renamed. This just requires including the parentId in the list where this function is called, in order to have the same effect.

Reply:
have resolved it with single function which adds the parent nodeid as well

const getChildNodeIds = (nodeId,subNodes = []) => {
	subNodes.push(nodeId);

	const childrenNodes = DAGRE_GRAPH.successors(nodeId);

	if (childrenNodes && childrenNodes.length) {

		childrenNodes.forEach((node) => {

			getChildNodeIds(node, subNodes);

		});
	}

	return subNodes;
};